File: benchmarks/bench_sprintf.js
#!/usr/bin/env node
// Node counterpart of bench_sprintf.strada. JS has no sprintf; the
// idiomatic equivalents (template literals, toFixed/toString(16),
// padStart/padEnd) produce the same output strings.
let t0 = Date.now();
let mlen = 0;
for (let i = 0; i < 500000; i++) {
    const line = `[INFO] user=u${i % 1000} req=${i} bytes=${(i * 37) % 100000} t=${((i % 500) / 7.0).toFixed(2)}ms`;
    mlen += line.length;
}
let t1 = Date.now();
console.log("mixed:", mlen, (t1 - t0) / 1000);

let nlen = 0;
for (let i = 0; i < 1000000; i++) nlen += `${i} ${i.toString(16)} ${i.toString(8)}`.length;
let t2 = Date.now();
console.log("numeric:", nlen, (t2 - t1) / 1000);

let flen = 0;
for (let i = 0; i < 500000; i++) {
    const v = i / 3.0;
    flen += `${v.toFixed(6)} ${v.toExponential(6)} ${String(v)}`.length;
}
let t3 = Date.now();
console.log("float:", flen, (t3 - t2) / 1000);

let wlen = 0;
for (let i = 0; i < 500000; i++) {
    const row = `${("row_" + (i % 100)).padEnd(20)}|${String(i).padStart(10)}|${String(i % 1000).padStart(8, "0")}|${((i % 1000) / 10.0).toFixed(1).padStart(6)}%`;
    wlen += row.length;
}
let t4 = Date.now();
console.log("width:", wlen, (t4 - t3) / 1000);
console.log("total:", (t4 - t0) / 1000);

File: benchmarks/bench_sprintf.pl
#!/usr/bin/env perl
# Perl counterpart of bench_sprintf.strada (identical workload).
use strict; use warnings; use Time::HiRes qw(time);

my $t0 = time;
my $mlen = 0;
for my $i (0..499999) {
    $mlen += length sprintf("[%s] user=%s req=%d bytes=%d t=%.2fms", "INFO", "u" . ($i % 1000), $i, ($i * 37) % 100000, ($i % 500) / 7.0);
}
my $t1 = time;
printf "mixed: %d %.6f\n", $mlen, $t1 - $t0;

my $nlen = 0;
$nlen += length sprintf("%d %x %o", $_, $_, $_) for 0..999999;
my $t2 = time;
printf "numeric: %d %.6f\n", $nlen, $t2 - $t1;

my $flen = 0;
for my $i (0..499999) { my $v = $i / 3.0; $flen += length sprintf("%.6f %e %g", $v, $v, $v) }
my $t3 = time;
printf "float: %d %.6f\n", $flen, $t3 - $t2;

my $wlen = 0;
for my $i (0..499999) {
    $wlen += length sprintf("%-20s|%10d|%08d|%6.1f%%", "row_" . ($i % 100), $i, $i % 1000, ($i % 1000) / 10.0);
}
my $t4 = time;
printf "width: %d %.6f\n", $wlen, $t4 - $t3;
printf "total: %.6f\n", $t4 - $t0;
